src/jit/codegenarmarch.cpp

```diff
@@ -1924,149 +1924,258 @@ void CodeGen::genCodeForCpBlkHelper(GenTreeBlk* cpBlkNode)
 }
 
 //----------------------------------------------------------------------------------
-// genCodeForCpBlkUnroll: Generates CpBlk code by performing a loop unroll
+// genCodeForInitBlkUnroll: Generate unrolled block initialization code.
 //
 // Arguments:
-//    cpBlkNode  -  Copy block node
+//    node - the GT_STORE_BLK or GT_STORE_OBJ node to generate code for
 //
-// Return Value:
-//    None
-//
-// Assumption:
-//  The size argument of the CpBlk node is a constant and <= CPBLK_UNROLL_LIMIT bytes.
-//
-void CodeGen::genCodeForCpBlkUnroll(GenTreeBlk* cpBlkNode)
+void CodeGen::genCodeForInitBlkUnroll(GenTreeBlk* node)
 {
-    // Make sure we got the arguments of the cpblk operation in the right registers
-    unsigned size    = cpBlkNode->Size();
-    GenTree* dstAddr = cpBlkNode->Addr();
-    GenTree* source  = cpBlkNode->Data();
-    GenTree* srcAddr = nullptr;
+    assert(node->OperIs(GT_STORE_BLK, GT_STORE_OBJ));
 
-    assert((size != 0) && (size <= CPBLK_UNROLL_LIMIT));
+#ifndef _TARGET_ARM64_
+    NYI_ARM("genCodeForInitBlkUnroll");
+#else
+    regNumber dstAddrBaseReg = genConsumeReg(node->Addr());
+    unsigned  dstOffset      = 0;
 
-    emitter* emit = GetEmitter();
+    regNumber srcReg;
+    GenTree*  src = node->Data();
 
-    if (dstAddr->isUsedFromReg())
+    if (src->OperIs(GT_INIT_VAL))
     {
-        genConsumeReg(dstAddr);
+        assert(src->isContained());
+        src = src->gtGetOp1();
     }
 
-    if (cpBlkNode->gtFlags & GTF_BLK_VOLATILE)
+    if (!src->isContained())
+    {
+        srcReg = genConsumeReg(src);
+    }
+    else
+    {
+        assert(src->IsIntegralConst(0));
+        srcReg = REG_ZR;
+    }
+
+    if (node->IsVolatile())
     {
-        // issue a full memory barrier before a volatile CpBlkUnroll operation
         instGen_MemoryBarrier();
     }
 
-    if (source->gtOper == GT_IND)
+    emitter* emit = GetEmitter();
+    unsigned size = node->GetLayout()->GetSize();
+
+    for (unsigned regSize = 2 * REGSIZE_BYTES; size >= regSize; size -= regSize, dstOffset += regSize)
+    {
+        emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, srcReg, srcReg, dstAddrBaseReg, dstOffset);
+    }
+
+    for (unsigned regSize = REGSIZE_BYTES; size > 0; size -= regSize, dstOffset += regSize)
     {
-        srcAddr = source->gtGetOp1();
-        if (srcAddr->isUsedFromReg())
+        while (regSize > size)
         {
-            genConsumeReg(srcAddr);
+            regSize /= 2;
+        }
+
+        instruction storeIns;
+
+        switch (regSize)
+        {
+            case 1:
+                storeIns = INS_strb;
+                break;
+            case 2:
+                storeIns = INS_strh;
+                break;
+            default:
+                storeIns = INS_str;
```

Comment:
Nit: could you please update this and the next switch with `case 4: ... default: unreached()` as discussed in https://github.com/dotnet/coreclr/pull/27035/files#r332302814?

Reply:
Sorry, missed that comment. I also forgot to bring some PR feedback commits from there. One moment...